cmd/kritis/gcbsigner/main.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/grafeas/kritis/pkg/kritis/crd/buildpolicy"
+	"github.com/grafeas/kritis/pkg/kritis/gcbsigner"
+	"github.com/grafeas/kritis/pkg/kritis/metadata/containeranalysis"
+	"github.com/grafeas/kritis/pkg/kritis/secrets"
+	"golang.org/x/net/context"
+
+	"cloud.google.com/go/pubsub"
+)
+
+func main() {
+	gcbProject := flag.String("gcb_project", "", "Id of the project running GCB")
+	gcbSubscription := flag.String("gcb_subscription", "build-signer", "Name of the GCB subscription")
+	resourceNamespace := flag.String("resource_namespace", os.Getenv("SIGNER_NAMESPACE"), "Namespace the signer CRDs and secrets are stored in")
+	flag.Parse()
+
+	err := run(*gcbProject, *gcbSubscription, *resourceNamespace)
+	if err != nil {
+		glog.Fatalf("Error running signer: %v", err)
+	}
+}
+
+func run(gcbProject string, gcbSubscription string, ns string) error {
+	ctx := context.Background()
+	client, err := pubsub.NewClient(ctx, gcbProject)
+	if err != nil {
+		return fmt.Errorf("Could not create pubsub Client: %v", err)
+	}
+
+	sub := client.Subscription(gcbSubscription)
+	for err == nil {
+		glog.Infof("Listening")
+		err = sub.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+			if err := process(ns, msg); err != nil {
+				glog.Errorf("Error signing: %v", err)
+				msg.Nack()
+			} else {
+				msg.Ack()
+			}
+		})
+	}
+	return fmt.Errorf("Error receiving message: %v", err)
+}
+
+func process(ns string, msg *pubsub.Message) error {
+	buildInfos, err := gcbsigner.ExtractImageBuildInfoFromEvent(msg)
+	if err != nil {
+		return err
```

Comment:
Add some context to the log message. process() has 4 paths that return errors, and I'm concerned that we won't be able to know which one created the message unless it gets annotated.

When the error is ultimately handled by the program's main function, it should provide a clear causal chain from the root problem to the overall failures reminiscent of a NASA accident investigation:

genesis: crashed: no parachute: G-switch failed: bad relay orientation

Context: https://golang.org/doc/effective_go.html#errors

Reply:
Done